### PR TITLE
[aes, prim] Encapsulate security-related synthesis attributes inside primitives 

### PR DESCRIPTION
**Pull request by vogelpi**
This is related to #4238.

We encapsulate security-related tool-specific synthesis attributes in primitives such that security-critical building blocks can also be safely synthesized with other tools by switching the primitives. This PR is a first step in that direction. To this end, two relevant additions to the primitives are made:
- A width parameter is added to the prim_buf cell. The default is set to one, so this shouldn't affect other designs using this primitive. The reason for adding this parameter is that it is heavily used inside the DOM S-Box. Placing loops around it as we did so far would make the code even harder to read.
- A new enable flip flop primitive is added. So far we just had a regular FF. Having no write enable can be very bad for SCA.

I've verified both functionality and security (formal) of the resulting DOM S-Box. What is left to do:
- [ ] Check FPGA implmentation.
- [x] Do the same for the masked Canright S-Boxes.

### CODE REVIEW

File: hw/ip/prim_generic/rtl/prim_generic_flop.sv
Comment (msfschaffner) [inline hw/ip/prim_generic/rtl/prim_generic_flop.sv:25]: nit: you could do `endmodule : prim_generic_flop` if you want...

**Comment by vogelpi** [comment]
@tjaychen @msfschaffner @eunchan this PR will break the nightly synthesis runs as the corresponding primitives need to be updated/created as well. I am aware that you guys are busy with silver. So don't worry about this one. It's not urgent. I just wanted to make you aware.

File: hw/ip/prim_generic/rtl/prim_generic_flop_en.sv
Comment (msfschaffner) [inline hw/ip/prim_generic/rtl/prim_generic_flop_en.sv:7]: Can you file an issue on me such that I do not forget adding the ASIC counterparts for this?

Thanks!

**Review by msfschaffner — APPROVED** [review]
LGTM, thanks!

**Comment by msfschaffner** [comment]
> @tjaychen @msfschaffner @eunchan this PR will break the nightly synthesis runs as the corresponding primitives need to be updated/created as well. I am aware that you guys are busy with silver. So don't worry about this one. It's not urgent. I just wanted to make you aware.

If the primitive does not exist, it will default to the generic implementation, which I think are all synthesizable on ASIC in this case. So this should not break our nightly synthesis.

File: hw/ip/prim_generic/rtl/prim_generic_buf.sv
Comment (cdgori) [inline hw/ip/prim_generic/rtl/prim_generic_buf.sv:8]: Is there any issue for the Width=1 case where we have changed this from a single-wire to a [0:0] bus?  I am not current on the behavior of VPI/DPI/PLI on these constructions (and especially how those interact with SV logic type), I just remember that there were some slight differences in how they were handled, even though they should be functionally the same.  I don't think this is an issue (or should block this PR), it just occurred to me when looking at the commit.

**Review by cdgori — APPROVED** [review]
LGTM.  My comment about [0:0] is probably inconsequential but I figured I should mention it.

Also, I read `aes_sbox_dom` carefully, but didn't check to make sure you removed all of the "keep" pragmas that were inline.  (I can check that all the ones you did remove are logically equivalent / correct, but not that you missed removing one...)

File: hw/ip/prim_generic/rtl/prim_generic_buf.sv
Comment (vogelpi) [reply hw/ip/prim_generic/rtl/prim_generic_buf.sv:8]: Oh yes, I didn't think about that. The only case in our code base I am aware of where the `[0:0]` makes a difference is around FPV. However, AFAIK this got fixed in the latest version of the FPV tool that we are now using (according to #5915). However, I am not aware about potential issues around VPI/DPI etc.

**Comment by vogelpi** [comment]
Thanks for your review Chris! I think I've caught all inline attributes. Grep doesn't find anything anymore.

Update: I've now also done the same for the masked Canright S-Boxes, re-ran functional verification, Yosys synthesis + formal analysis with the same results: the no-reuse version passes stable verification but fails in transient mode, the normal version fails always.

I suggest to merge this once CI passes. Verification on FPGA will happen later. I'll need to debug the SCA platform first but before that some more changes will happen in the chip-level.

File: hw/ip/prim_generic/rtl/prim_generic_buf.sv
Comment (cdgori) [reply hw/ip/prim_generic/rtl/prim_generic_buf.sv:8]: As I said, I wouldn't block this PR on the change, I just know that net and single-wire bus sometimes behaved slightly differently historically.  (DFT insertion tools were famous for creating these single-wire bus things, I seem to recall.)

**Comment by cdgori** [comment]
> I suggest to merge this once CI passes. Verification on FPGA will happen later. I'll need to debug the SCA platform first but before that some more changes will happen in the chip-level.

This seems reasonable to me.

File: hw/ip/prim_generic/rtl/prim_generic_flop_en.sv
Comment (vogelpi) [reply hw/ip/prim_generic/rtl/prim_generic_flop_en.sv:7]: Yep, it's here #6097.

**Pull request merged**
